Redesign mount propagation

jsafrane opened the pull request:
The proposal won't work as it was merged, it makes too many directories as shared (see #648).

A different approach is needed, I've chosen 'Add an option in VolumeMount API', but I would be fine also with 'Add an option in HostPathVolumeSource', there is only very small difference to me.

The proposal also describes how it will be implemented, especially during alpha phase.

Fixes #648

@kubernetes/sig-node-proposals @kubernetes/sig-storage-proposals

jsafrane commented:
For alpha annotation, I know there is attempt to have real fields in alpha in https://docs.google.com/document/d/1wuoSqHkeT51mQQ7dIFhUKrdi3-1wbKrNWeIL4cKb9zU/edit#, but it seems it has not been implemented yet. Alpha annotation is still described as preferred way to introduce new fields in https://github.com/kubernetes/community/blob/master/contributors/devel/api_changes.md

yujuhong commented:
Adding a couple of folks who were on the original proposal.

/cc @vishh @euank

dchen1107 commented:
We discussed at sig-node meeting, and agreed with the redesign the proposal scope. But this feature is not targeted for 1.7 release.

inc0 commented:
This is required change for openstack-on-kubernetes too. Could we please add sig-openstack to this review?

kfox1111 commented:
+1 for sig-openstack.

I would also greatly benefit from it for some high energy physics work I do.

If the patch set is ready for 1.7 will it go in?

Is there anything I can do to help get it ready faster?

ddysher commented:
This is also required for local volume provisioner.  The provisioner is running as a daemonset to monitor local volume and post PV to kubernetes.  To correctly detect volume capacity, the daemonset pod must mount hostpath (path where volume is detected) as rshared.

Are we targeting alpha in 1.8 now?

jsafrane commented:
yes, we missed 1.7

ddysher commented:
Thanks for confirm.  I'd be happy to test it out when implementation is out.

jsafrane commented:
I updated the PR with the latest development:

- added links to Kubernetes PRs (waiting for review)
- added note that the default propagation should be `rslave`. This is what was agreed in the previous proposal, now it's more explicit.
- I tested my PRs with various distros and I stumbled upon Google COS (former GCI). It runs Docker in its own mount namespace with slave mount propagation. I.e. it can't ever run a pod with shared mount propagation. There should be a check in kubelet that tests if it can run a pod with shared mount propagation, at least on `/var/lib/kubelet`, and refuse to start if it can't.
  - This test is quite tricky, the best would be to run some dummy container (`gcr.io/*/pause`?) with `docker run -v /var/lib/kubelet:/tmp:rshared` and check if it failed.
  - Note that COS-60 fixes this issue, however the test might be useful if someone runs another old or ~~weird~~ unsupported distro.

msau42 commented on contributors/design-proposals/propagation.md:
```diff
@@ -142,13 +153,41 @@ distros.
+  it will be marked as deprecated from the beginning.
+  Developers / testers can enable it in their clusters manually.
+  Mount propagation may be redesigned or even removed in any future release.
+* The default mount propagation will be `rslave`, which is different to current
```
Earlier, it says the default is `private`, to not break backwards compatiblity. Why is it `rslave` here?

jsafrane replied:
thanks for spotting this, the default should be rslave, which *should* not break backward compatibility (testing needed!) and enables this use case https://github.com/kubernetes/kubernetes/pull/46444#issuecomment-304439683

kfox1111 commented:
Ping. What is the status of getting this into 1.8? I don't see it marked for 1.8.

I'm stuck on k8s 1.4 currently, as I need this kind of functionality and docker 1.9 is the last version to default mounts to shared and k8s 1.4 is the last version to support docker 1.9.

I would be happy if this was merged but hidden behind an experimental flag or something, just so those of us that need to make forward progress can do so.

kfox1111 commented:
ping

luxas commented:
@jsafrane It would indeed be nice to have this as alpha in v1.8

jsafrane commented:
@luxas, all that's missing is review of #46444

saad-ali commented:
/lgtm
/approve 

We discussed this design in a meeting on Monday July 24. I am fine with this.

One thing I just remembered is that there is a pending design to add "HostPathType" to hostpath volumesource: https://github.com/kubernetes/kubernetes/pull/46597
Let's make sure that these two options (propagation and hostPathType play well together)

CC @timstclair @dixudx

dixudx commented:
@saad-ali `propagation` will not conflict with `HostPathType`.

jsafrane commented:
@timstclair was there any discussion about feature gate vs --exeprimental cmdline argument? I am happy with both, I just need a decision.

tallclair commented on contributors/design-proposals/propagation.md:
```diff
@@ -142,13 +153,41 @@ distros.
-non-privileged', an environment check and an WARNING log will be emitted about
-whether propagation mode is supported.
-
+* We will take 'Add an option in VolumeMount API' (with an annotation during
```
I disagree with this decision. I think this should be added to the HostPathVolumeSource, since it only applies to HostPath volumes. I think it's a relatively small edge case that there are multiple containers in the same pod mounting the same HostPath volume, where one wants it to be shared and the other wants slave. In that case, if one of the containers was not privileged, it would get the non-shared mount anyway.

jsafrane replied:
There were people who wound like to run EmptyDir with shared mount propagation so a pod could have two containers, one privileged one that mounts something into the EmptyDir so the second container can consume it. https://github.com/kubernetes/kubernetes/pull/46444#issuecomment-310857111

IMO it's an useful use case and we could implement it when going out of alpha.

tallclair replied:
There are several fields [1] that apply to multiple VolumeSources, and the fields are replicated amongst the applicable sources. On the other hand, all the fields in VolumeMount are universally applicable to all volumes. I think it would be more consistent in this case to just include the propagation mode in both the HostPath and EmptyDir VolumeSources.

[1] FSType, ReadOnly, DefaultMode

jsafrane replied:
Ok, that makes sense. On the other hand, with `MountPropagation` as `HostPathVolumeSource` field we can't validate if container is privileged when running with a HostPath with shared mount propagation - propagation is in a PersistentVolume object and privileged mode is set in Pod and validator sees only one of these objects. Last time I checked we did not do cross-object validation.

tallclair replied:
So it's a bit weird, but my proposal is that it should be OK to have a shared hostpath volume mounted into an unprivileged container, but that container doesn't get shared (it gets slave). This let's you have the state where there are 2 containers in a pod, and only one gets the shared propagation. I suppose you could validate that if it's set to shared, there must be at least 1 privileged container, but I think it's ok without that. We should make sure this behavior is documented on the field though.

tallclair commented on contributors/design-proposals/propagation.md:
```diff
@@ -142,13 +153,41 @@ distros.
+  `/var/lib/kubelet` and it can be shared into containers.
+  E.g. Google COS-58 runs Docker in a separate mount namespace with slave
+  propagation and thus can't run a container with shared propagation on
+  anything. Other container engines should follow the suit.
```
I'm confused by this. Isn't this saying that containers with shared mounts cannot be run on COS, which would prevent the Kubelet from starting under this new model?

jsafrane replied:
COS 58 runs docker 1.12 and its systemd service runs docker daemon it its own mount namespace with slave mount propagation from the host. So a container can never export anything to the host, only to the docker mount namespace. This is fixed in COS 60 with docker 1.13.

tallclair replied:
I'm just confused by the `Other container engines should follow the suit.` line... I think what you're trying to say is:
```
Google COS-60 and above support shared propagation, and other Kubernetes node OSs should follow suit.
```

jsafrane replied:
I have very limited knowledge about rkt and its support for mount propagation. It should follow Docker's model.

thockin replied:
I am confused by this too.  Does this mean that this model can share mounts between docker containers (shared would propagate up to docker's mountns) but not up to the host OS?

jsafrane replied:
Yes, containers can shared mounts among each other (and to docker mount NS), but kubelet won't see these mounts. Fixed in Docker 1.13 and I think 1.12 is going to be deprecated when Kubernetes 1.8 is released.

tallclair commented on contributors/design-proposals/propagation.md:
```diff
@@ -142,13 +153,41 @@ distros.
+* Node conformance suite will check that mount propagation in /var/lib/kubelet
+  works.
+* During alpha, all the behavior above must be explicitly enabled by
+  `kubelet --experimental-enable-mount-propagation`
```
Use a feature gate instead, with a new "experimental" release track (https://github.com/kubernetes/kubernetes/blob/9af1ff3dce1f71f31c051b5b84b6c187ebdc6476/staging/src/k8s.io/apiserver/pkg/util/feature/feature_gate.go#L64)

jsafrane replied:
ok, switched to feature gate

liggitt commented on contributors/design-proposals/propagation.md:
```diff
@@ -142,13 +153,41 @@ distros.
-whether propagation mode is supported.
-
+* We will take 'Add an option in VolumeMount API' (with an annotation during
+  alpha instead of real VolumeMount field):
```
-1 on alpha annotation, feature-gated API field is preferred

jsafrane replied:
I do know about https://docs.google.com/document/d/1wuoSqHkeT51mQQ7dIFhUKrdi3-1wbKrNWeIL4cKb9zU/, but it seems to me as it's work in progress.

* I haven't noticed any announce of new API style on kubernetes-devel

* https://github.com/kubernetes/community/blob/master/contributors/devel/api_changes.md#adding-unstable-features-to-stable-versions still recommends alpha fields for unstable features.

* Nobody uses `// +gate` comment in any `types.go` in Kubernetes

jsafrane replied:
Discussed with @liggitt elsewhere, I'm dropping the annotation and real field with a feature gate will be used.

jsafrane commented:
Just to make it clear, if we go with `HostPathVolumeSource`:

```go
type HostPathVolumeSource struct {
	Path string `json:"path"`
	// Mount the host path with propagation mode specified.
        Propagation PropagationMode `json:"propagation,omitempty"`
}
```

`HostPathVolumeSource.Propagation` is **string** and not **pointer to a string**, right? It's empty when omitted. Should it have a default value if it's alpha?

jsafrane commented:
Pushed new version

- `struct Propagation` renamed to `MountPropagation`, who knows what other kinds of propagation will be there
- moved to `HostPathVolumeSource`

I hope it's the final version.

tallclair commented:
/lgtm

thockin commented on contributors/design-proposals/propagation.md:
````diff
@@ -89,18 +100,23 @@ The new `HostPathVolumeSource` will look like:
 }
 ```
 
+The default mount propagation is `rslave`. Any HostPath can ask for `private`.
````
Why would anyone ask for private?  What value does the extra surface area provide?

jsafrane replied:
Dunno, it's just safety hatch when something breaks. Maybe a pod that uses whole "/" as HostPath may not want to get further mounts from the host.

thockin replied:
I don't think we should design to hypotheticals, especially when it involves complex API surface.

bgrant0607 replied:
Is there a compelling reason why this couldn't be a simple boolean?

kfox1111 replied:
because there are three options at present. rslave, rshared, and private.

thockin commented on contributors/design-proposals/propagation.md:
```diff
@@ -73,13 +79,18 @@ type VolumeMount struct {
 Opinion against this:
 
-1. This will affect all volumes, while only HostPath need this.
+1. This will affect all volumes, while only HostPath need this. It could be
```
This is a really nasty implication.

kfox1111 replied:
I disagree only hostpath needs propagation. I totally could see it with at least emtpyDir too. Say I want to have a cvmfs sidecar to my main pod? I could use cvmfs with its automounter to mount cvmfs sumbmounts in the emptydir and expose the whole /cvmfs automounter to the other containers in the pod.

I still would really like to see a namedEmptyDir implementation that could use it too.

thockin replied:
This is an interesting use case in its own, and if we want to consider emptyDir with propagation, we need to pay extra attention to cleanup.  I have observed that this can lead to "leaked" mounts.  It definitely changes the calculus on this approach.

What is namedEmptyDir ?

bgrant0607 replied:
It doesn't have to be the case that adding the option to VolumeMount would imply that it would have to be supported for all volumes. The supported volumes could be documented on the field and whitelisted in validation.

thockin commented on contributors/design-proposals/propagation.md:
```diff
@@ -109,6 +125,12 @@ Opinion against this:
 
 1. (From @jonboulle) May cause cross-runtime compatibility issue.
 
+1. It's not possible to validate a pod + mount propagation. Mount propagation
```
The validators are just code - we can change this, or we can add another validation at the end of ValidatePodSpec() that cross-checks privileged concerns, right?

jsafrane replied:
At the time of validation of a pod the PVCs it uses may not be bound to a PVs yet, so there is nothing to validate. Binding to PVs can happen later.

thockin replied:
Ahh, sorry, I missed that this was about PV.  We can reject at runtime, if we pursue this.  That said, HostPath as PV is still a terrible idea, and nobody should use it.

kfox1111 replied:
isn't local volumes using it? pv generated with hostpath and injected attribute to pin it to a particular host?

jsafrane replied:
> HostPath as PV is still a terrible idea, and nobody should use it.

+1000

> isn't local volumes using it?

Local has its own volume type named `Local` (`LocalVolumeSource`)
￼

kfox1111 replied:
oh. I didn't realize that. Does that mean the cool annotation thing only works with local volumes?

That would be a bummer. as the annotation could be useful in other ways, such as tagging a cinder based pv so that the pod it gets attached to only lands on vm's when you have a hybrid openstack/baremetal cluster (I do)

msau42 replied:
The PV node affinity is designed so that it could be used on other volume types, but right now it is behind the local volume feature gate.

msau42 replied:
But even then, it will not work out of the box for hostpath PVs.  There are limitations in how the hostpath volume plugin is implemented that prevent it from working in a multi-node cluster.

kfox1111 replied:
@thockin I've been toying around with the idea of contributing a 'namedEmptyDir' volume type. what it does, is similar to an emptyDir, but it is given a name unique to a namespace, and its lifecycle doesn't have it deleted until the last pod on the same host that references it goes away.
So, it would be stored somwhere like /var/lib/kubelet/volumes/namedEmptydir/<namespace name>/<name from user>/ (or whatever)

This would allow multiple k8s objects such as deamonsets or daemonsets/deployments/etc to coexist on the host and share some transient storage amongst them but without having to book keep where to store them using a hostpath and ensuring different name spaces don't stop on each others data.

Its use works in concert with this proposal in that say you have a fuse based system such as cvmfs that has a backing cache. You could have several different cvmfs daemonsets or pods consume the cache dir from a namedEmptyDir on the host via "cvmfs-cache" and have another one called "cvmfs" with mount propagation set to shared where cvmfs automounter is running.  Any pod in the namespace can then mount the namedEmptyDir 'cvmfs' and the the automount cvmfs volume would work optimally.

Ideally for security, you would have the volume mounted as shared on the autofs fuse filesystem container and have it slave on the things consuming it. Even more reason for having it as a flag and not a volumetype.

thockin commented on contributors/design-proposals/propagation.md:
```diff
@@ -142,13 +164,38 @@ distros.
+* We will take 'Add an option in HostPathVolumeSource API'
+  * With an alpha feature gate in 1.8.
+  * Only privileged containers can use `rshared` mount propagation.
+    * When non-privileged container uses `rshared` HostPath, it silently
```
silent failure is bad - we should just fail

jsafrane replied:
Do you mean setting `PodStatus.Phase: Failed`? Well... I can do it (if someone shows me the right place in kubelet), however it looks too offensive to me.

thockin replied:
We need to define some sane behavior.  Kubelet could reject it, or just leave it pending with an event

jsafrane replied:
`Pending` + event is easy, I am not sure about rejection. Does it mean pod admission? Are all PVCs bound at this time? @tallclair @derekwaynecarr

tallclair replied:
The modivation for the silent downgrade is to cover the case where you have a Pod with a shared hostpath volume shared by 2 containers, 1 privileged that does the mounting, and 1 non-privileged that simply reads from the mounted volume. Since we decided to move the mount propagation setting into the volume definition (instead of the volume mount). I don't like making the behavior conditional on privileged either, but I think I prefer it to having the setting in the VolumeMount (since it only applies to a small subset of volumes).

thockin replied:
I don't understand the use case.  Does it not work?

jsafrane replied:
Consider a pod that has two containers, one privileged "sidecar" and one unprivileged mysql. The sidecar one just mounts some volume for the mysql and unmounts it when the pod dies. If I add admission handler to kubelet, it will see a pod with an unprivileged container using a HostPath ( EmptyDir would be better in this case) with rshared propagation and it will reject it.

jsafrane commented:
> What if we defined a whole new volume type that was abstractly "a place to mount volumes" - and we handle it how we need in kubelet

I thought that's what `struct VolumeSource` is for.

On sig-storage, we have embargo on new volume plugins and creating a new one just to expose extra Linux flags does not sound good to me. I would not expect that PVs are portable at all, we put too much Linux stuff to them already - e.g. mount options.

kfox1111 commented:
Uh, last I looked the distro's were only shipping 1.12 still?

jsafrane commented:
@kfox1111 what distros? I know RHEL/CentOS 7 and COS (or GCI) < 60 ships 1.12. COS 59 is going to die soon-ish and I am working on RHEL 7.

kfox1111 commented:
el7 in particular I guess. I didn't think redhat was going to support 1.13? You probably know more about that then I do though.

thockin commented:
re: embargo on new volumes.  It has to be more nuanced than a blanket embargo.  Some volumes don't make sense being built-in, and some just do.  In this case, we're exposing a pretty complex API surface, with weird corner cases and failure modes, that barely we can really understand just to avoid a new volume?  No.

I would like to understand what the pros/cons of a new volume type would be.  Would it prevent anything that we know or strongly suspect we will need?  Would it need to be exposed as PV at all (I think not)?  Does it simplify the impl or make it more complex or neutral?  Is it easier or harder to document?

It is not my intent to drag my feet, but this is sort of important.

@saad-ali @jingxu97

kfox1111 commented:
@thockin Like I said above, I can see other volumes that may want to share the same functionality. I think adding another volume type would be a step in the wrong direction as it couldn't be shared between the other volume types.

kfox1111 commented:
Also, several different groups are really hamstrung by the lack of this functionality. Can we please get it into 1.8 as alpha while we explore the way to do it 100% proper in beta in a following release? I'm currently stuck on k8s 1.4 because of this.

thockin commented:
I am sorry that efforts are hamstrung, but this is real API surface.

bgrant0607 commented on contributors/design-proposals/propagation.md:
````diff
@@ -65,6 +65,12 @@ and references to network namespaces persist.
 ```go
+const (
+	MountPropagationRShared  MountPropagationMode = "RShared"
+	MountPropagationRSlave   MountPropagationMode = "RSlave"
````
Abbreviation is discouraged.

jsafrane replied:
`MountPropagationRecursiveShared` sounds really ugly and it's not used nowhere in kernel / docker  documentation. Renamed to `Shared` with a note that it's applied recursively (imo it's still quite confusing...)

jsafrane commented:
> I would like to understand what the pros/cons of a new volume type would be.

So, let there be `LinuxVolumeSource`:

```go
type LinuxVolume struct {
    MountPropagation string
    WrappedVolume *VolumeSource
}

type VolumeSource struct {
   ... aws, gce, hostPath, emptyDir, git, ...
    LinuxVolume *LinuxVolumeSource
}
```

1. It's recursive. Recursive structures in API are IMO ugly.
2. Does not help with application portability at all. If an application needed mount propagation it won't work on macOS / Windows at all, regardless if `LinuxVolume` or `HostPath` was used.
3. I do not see any added value other than increased complexity and confusion.

In my world, `VolumeSource` describes *where* to get the data from. I still personally think that it should not include things like `readOnly` or `mountPropagation`. It should be in `VolumeMount struct` which says *how* the data should be made available. It's much easier to validate it, it goes well along with `VolumeMount.readOnly`. I can accept that `HostPathVolumeSource` and `EmptyDirVolumeSource` is the right place, it will work well for all use cases that I know about. I don't see any reason why to invent a new volume type.

bgrant0607 commented:
For reference:
https://lwn.net/Articles/689856/
https://www.kernel.org/doc/Documentation/filesystems/sharedsubtree.txt

kfox1111 commented:
I agree. Properties on how something gets mounted is a totally different thing than what the thing storing the data is. hostPath, emptyDir, etc will need the flag. not, recreate volume types for rsharedHostPath and rsharedEmptyDir. Once namedEmptyDir becomes a thing, it will want it too.

kfox1111 commented:
@thockin yeah, I get the desire for a correct api. Totally a good thing to get right. Just saying, for an alpha, getting it perfect is not so critical. I'm totally fine accepting the risk of writing software against an api that is expected to drastically change in future releases. That part is easy. Living without this feature is really really hard. I'm facing a 1.4 -> 1.8 migration path already, with broken steps in between. :/ Several of these releases are already unsupported.

As for openstack on k8s, it can't really be a complete thing until this feature merges in some form or another. Neutron really needs it.

bgrant0607 commented on contributors/design-proposals/propagation.md:
```diff
@@ -142,13 +164,38 @@ distros.
+* Node conformance suite will check that mount propagation in /var/lib/kubelet
+  works.
+* During alpha, all the behavior above must be explicitly enabled by
+  `kubelet --feature-gates=MountPropagation=true`
```
ICYMI

Field gate proposal: https://github.com/kubernetes/community/pull/869

jsafrane replied:
Thanks!

bgrant0607 commented on contributors/design-proposals/propagation.md:
```diff
@@ -73,13 +79,18 @@ type VolumeMount struct {
 	MountPath string `json:"mountPath"`
 	// Optional.
-	Propagation string `json:"propagation"`
+	Propagation PropagationMode `json:"propagation,omitempty"`
```
This field name is very vague. So is the file name, by the way. Propagation of what? In what direction?

jsafrane replied:
> This field name is very vague... Propagation of what? In what direction?

Renamed to `MountPropagation` with more documentation around. Still, I don't want to put whole kernel documentation there.

> So is the file name

What file name?

bgrant0607 replied:
The proposal filename.

thockin commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/community/659)
<!-- Reviewable:end -->

jsafrane commented:
We had a productive meeting yesterday with @thockin and @saad-ali and we're back at "Add an option in VolumeMount API".

* While we do not see any use case for any other volume than HostPath or EmptyDir to be `rshared`, we will not block it, at least not in alpha. It must be explicitly enabled and having rshared GCE volume is not more harmful than HostPath.

* Renamed `(r)shared` and `(r)slave` in the API to `Bidirectional` and `HostToContainer`.

* Removed `private` as we do not see any use case for it.

* We use enum instead of simple bool to be able to add `private` and non-recursive `shared` and `slave` if we need so. We can't extend `bool` so easily.

I did more edits describing what exactly will be in alpha, see the last commit.

@bgrant0607, can you please look at the API again? PR with Kubernetes changes will follow very shortly to catch 1.8

saad-ali commented:
Revised proposal:

/lgtm
/approve

Thanks @jsafrane

saad-ali commented:
/lgtm

bgrant0607 commented:
Github notifications are useless :-(
